tiledb/api/src/filter.rs

```diff
@@ -82,445 +415,185 @@ impl<'ctx> Filter<'ctx> {
-            &mut c_level as *mut std::ffi::c_int as *mut std::ffi::c_void,
-        )
-        .map(|_| c_level as i32)
-    }
-
-    pub fn get_compression_reinterpret_datatype(
-        &self,
-    ) -> TileDBResult<ffi::Datatype> {
-        let mut c_fmt: std::ffi::c_uchar = 0;
-        let res = self.get_option(
-            ffi::FilterOption::COMPRESSION_REINTERPRET_DATATYPE,
-            &mut c_fmt as *mut std::ffi::c_uchar as *mut std::ffi::c_void,
-        );
-        match res {
-            Ok(()) => match ffi::Datatype::from_u32(c_fmt as u32) {
-                Some(dtype) => Ok(dtype),
-                None => Err(Error::from("Invalid compression reinterpret datatype returned from core."))
-            },
-            Err(msg) => Err(msg),
+        if res != ffi::TILEDB_OK {
+            return Err(self.context.expect_last_error());
         }
-    }
 
-    pub fn get_float_bytewidth(&self) -> TileDBResult<u64> {
-        let mut c_width: std::ffi::c_ulonglong = 0;
-        self.get_option(
-            ffi::FilterOption::SCALE_FLOAT_BYTEWIDTH,
-            &mut c_width as *mut std::ffi::c_ulonglong as *mut std::ffi::c_void,
-        )
-        .map(|_| c_width as u64)
-    }
-
-    pub fn get_float_factor(&self) -> TileDBResult<f64> {
-        let mut c_factor: std::ffi::c_double = 0.0;
-        self.get_option(
-            ffi::FilterOption::SCALE_FLOAT_FACTOR,
-            &mut c_factor as *mut std::ffi::c_double as *mut std::ffi::c_void,
-        )
-        .map(|_| c_factor as f64)
-    }
-
-    pub fn get_float_offset(&self) -> TileDBResult<f64> {
-        let mut c_factor: std::ffi::c_double = 0.0;
-        self.get_option(
-            ffi::FilterOption::SCALE_FLOAT_OFFSET,
-            &mut c_factor as *mut std::ffi::c_double as *mut std::ffi::c_void,
-        )
-        .map(|_| c_factor as f64)
-    }
-
-    pub fn get_positive_delta_max_window(&self) -> TileDBResult<u32> {
-        let mut c_width: std::ffi::c_uint = 0;
-        self.get_option(
-            ffi::FilterOption::POSITIVE_DELTA_MAX_WINDOW,
-            &mut c_width as *mut std::ffi::c_uint as *mut std::ffi::c_void,
-        )
-        .map(|_| c_width as u32)
-    }
-
-    pub fn get_webp_input_format(
-        &self,
-    ) -> TileDBResult<ffi::WebPFilterInputFormat> {
-        let mut c_fmt: std::ffi::c_uchar = 0;
-        let res = self.get_option(
-            ffi::FilterOption::WEBP_INPUT_FORMAT,
-            &mut c_fmt as *mut std::ffi::c_uchar as *mut std::ffi::c_void,
-        );
-        match res {
-            Ok(()) => {
-                match ffi::WebPFilterInputFormat::from_u32(c_fmt as u32) {
-                    Some(fmt) => Ok(fmt),
-                    None => Err(Error::from(
-                        "Invalid WebP input filter format returned from core.",
-                    )),
-                }
+        let get_compression_data =
+            |compression_type| -> TileDBResult<FilterData> {
```

Comment:
Style nit: I imagine this would be a whole lot easier to read if we renamed the argument to `kind` and then assigned the level and reinterpret type to temporaries so that the end is just:

```rs
Ok(FilterData::Compression(CompressionData { kind, level, reinterpret_datatype }))
```